rules.go

```diff
@@ -0,0 +1,482 @@
+	whitelistExcludeOptionsRegex map[string]*regexp.Regexp
+	blacklistIncludeOptions      map[string][]*ruleAdBlock
+	blacklistIncludeOptionsRegex map[string]*regexp.Regexp
+	blacklistExcludeOptions      map[string][]*ruleAdBlock
+	blacklistExcludeOptionsRegex map[string]*regexp.Regexp
+}
+
+func matchWhite(ruleSet RuleSet, req Request) bool {
+	if ruleSet.whitelistRegex != nil && ruleSet.whitelistRegex.MatchString(req.URL.String()) {
+		return true
+	}
+
+	options := extractOptionsFromRequest(req)
+	for option, active := range options {
+		if ruleSet.whitelistIncludeOptionsRegex[option] != nil && ruleSet.whitelistIncludeOptionsRegex[option].MatchString(req.URL.String()) {
+			return active == true
+		}
+		if ruleSet.whitelistExcludeOptionsRegex[option] != nil && ruleSet.whitelistExcludeOptionsRegex[option].MatchString(req.URL.String()) {
+			return active == false
+		}
+	}
+	return false
+}
+
+func matchBlack(ruleSet RuleSet, req Request) bool {
+	if ruleSet.blacklistRegex != nil && ruleSet.blacklistRegex.MatchString(req.URL.String()) {
+		return true
+	}
+
+	rulesIncluded := []*ruleAdBlock{}
+	for _, rule := range ruleSet.blacklistDomainsNoOptions {
+		include := true
+		matched := false
+		for domain, allowed := range rule.domains {
+			if strings.Contains(req.URL.Hostname(), domain) {
+				include = include && allowed
+				matched = true
+			}
+		}
+		if matched && include {
+			rulesIncluded = append(rulesIncluded, rule)
+		}
+	}
+
+	blacklistDomainsRegex := CombinedRegex(rulesIncluded)
+	if blacklistDomainsRegex != nil && blacklistDomainsRegex.MatchString(req.URL.String()) {
+		return true
+	}
+
+	options := extractOptionsFromRequest(req)
+	for option, active := range options {
+		if ruleSet.blacklistIncludeOptionsRegex[option] != nil && ruleSet.blacklistIncludeOptionsRegex[option].MatchString(req.URL.String()) {
+			return active == true
+		}
+		if ruleSet.blacklistExcludeOptionsRegex[option] != nil && ruleSet.blacklistExcludeOptionsRegex[option].MatchString(req.URL.String()) {
+			return active == false
+		}
+	}
+	return false
+}
+
+func (ruleSet *RuleSet) Allow(req Request) bool {
+	if ok := matchWhite(*ruleSet, req); ok {
+		return true
+	}
+	if ok := matchBlack(*ruleSet, req); ok {
+		return false
+	}
+	return true
+}
+
+func readLines(path string) ([]string, error) {
+	f, err := os.Open(path)
+
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	reader := bufio.NewReader(f)
+	lines := []string{}
+	for line := []byte{}; err == nil; line, _, err = reader.ReadLine() {
+		sl := strings.TrimSuffix(string(line), "\n\r")
+		if len(sl) == 0 {
+			continue
+		}
+		lines = append(lines, sl)
+	}
+
+	return lines, nil
+}
+
+func NewRulesSetFromFile(path string) (*RuleSet, error) {
+	lines, err := readLines(path)
+	if err != nil {
+		return nil, err
+	}
+	return NewRuleSetFromStr(lines)
+}
+
+func NewRuleSetFromStr(rulesStr []string) (*RuleSet, error) {
+	logger.Init("NewRuleSetFromStr", true, true, ioutil.Discard)
```

Comment:
This library should have no logger! Let's imagine a situation when a user uses its own logger to log JSON etc. And now we see some plain text messages with the incorrect format in the stream.

Please remove logger